### ISSUE TITLE
use of extern in accessorClasses.h

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#218]
Here's the replication of the CRAN error (albeit with an extra warning about the ";").  However if I take out the externs I get a different error when compiling accessorClasses.cpp because of duplication between accessorClasses.cpp and accessorClasses.h (see bottom). Then if I remove the template line from accessorClasses.cpp, I get the error at the very bottom...

To be able to see these errors, I believe that inserting 
CXXFLAGS=-std=c++98 -mtune=native -g -O2 -Wall -pedantic
in your $HOME/.R/Makevars will reveal them.

g++ -I/usr/share/R/include -DNDEBUG -I../inst/include      -fpic  -std=c++98 -mtune=native -g -O2 -Wall -pedantic -c nimble.cpp -o nimble.o
In file included from nimble.cpp:3:0:
../inst/include/nimble/accessorClasses.h:78:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h:85:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h:89:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h:98:150: warning: ISO C++ 1998 forbids the use of ‘extern’ on explicit instantiations [-Wpedantic]
 extern template NimArr<1, double> getParam_1D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                      ^
../inst/include/nimble/accessorClasses.h:99:156: warning: ISO C++ 1998 forbids the use of ‘extern’ on explicit instantiations [-Wpedantic]
 extern template NimArr<1, double> getParam_1D_double<NimArr<1, double> >(const NimArr<1, double> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                            ^
../inst/include/nimble/accessorClasses.h:105:150: warning: ISO C++ 1998 forbids the use of ‘extern’ on explicit instantiations [-Wpedantic]
 extern template NimArr<2, double> getParam_2D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                      ^
../inst/include/nimble/accessorClasses.h:106:156: warning: ISO C++ 1998 forbids the use of ‘extern’ on explicit instantiations [-Wpedantic]
 extern template NimArr<2, double> getParam_2D_double<NimArr<1, double> >(const NimArr<1, double> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                            ^
g++ -shared -L/usr/lib/R/lib -Wl,-z,relro -o nimble.so nimble.o -L../inst/CppCode -lnimble -Wl,-rpath=/usr/local/lib/R/site-library/nimble/CppCode -llapack -lblas -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/nimble/libs

==========================
error if externs removed
===========================

g++ -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-NmmxMb/r-base-3.3.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c accessorClasses.cpp -o accessorClasses.o
accessorClasses.cpp:33:143: error: duplicate explicit instantiation of 'NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, int>]' [-fpermissive]
 template NimArr<1, double> getParam_1D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                               ^
accessorClasses.cpp:34:149: error: duplicate explicit instantiation of 'NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, double>]' [-fpermissive]
 template NimArr<1, double> getParam_1D_double<NimArr<1, double> >(const NimArr<1, double> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                     ^
accessorClasses.cpp:46:143: error: duplicate explicit instantiation of 'NimArr<2, double> getParam_2D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, int>]' [-fpermissive]
 template NimArr<2, double> getParam_2D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                               ^
accessorClasses.cpp:47:149: error: duplicate explicit instantiation of 'NimArr<2, double> getParam_2D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, double>]' [-fpermissive]
 template NimArr<2, double> getParam_2D_double<NimArr<1, double> >(const NimArr<1, double> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                   
================
after removing duplicate lines such as this from accessorClasses.cpp 
template NimArr<1, double> getParam_1D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
then I get the following error.
==================

g++ -I/usr/share/R/include -DNDEBUG -I../inst/include      -fpic  -std=c++98 -mtune=native -g -O2 -Wall -pedantic -c nimble.cpp -o nimble.o
In file included from nimble.cpp:3:0:
../inst/include/nimble/accessorClasses.h:78:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h:85:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h:89:2: warning: extra ‘;’ [-Wpedantic]
 };
  ^
../inst/include/nimble/accessorClasses.h: In instantiation of ‘NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, int>]’:
../inst/include/nimble/accessorClasses.h:98:143:   required from here
../inst/include/nimble/accessorClasses.h:98:143: error: explicit instantiation of NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, int>]’ but no definition available [-fpermissive]
 template NimArr<1, double> getParam_1D_double<NimArr<1, int> >(const NimArr<1, int> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                               ^
../inst/include/nimble/accessorClasses.h: In instantiation of ‘NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, double>]’:
../inst/include/nimble/accessorClasses.h:99:149:   required from here
../inst/include/nimble/accessorClasses.h:99:149: error: explicit instantiation of NimArr<1, double> getParam_1D_double(const paramIDtype&, const oneNodeUseInfo&, int) [with paramIDtype = NimArr<1, double>]’ but no definition available [-fpermissiv]
 template NimArr<1, double> getParam_1D_double<NimArr<1, double> >(const NimArr<1, double> &paramID, const oneNodeUseInfo &useInfo, int iNodeFunction);
                                                                                                                                                     ^

**Comment by paciorek** [comment]
I was able to get rid of the complaints about the 'extern' by removing the entire 'extern template' code lines from accessorClasses.h.  In his email Perry mentioned something about these being in the .h to reduce redundant specializations so I'm suspecting this is fine, but perhaps a bit inefficient. I think we can close this once Perry signs off.

**Comment by perrydv** [comment]
I am not sure that is the right solution.  I have a branch where instead I put the full template definitions in the .h and removed any attempt at explicit specializations.  I ran getParam tests, which should be the relevant ones, and it seemed to pass.  Will do pull request later.

**Comment by paciorek** [comment]
are these changes in extern_issues?  I'd like to do a full suite of tests.

**Comment by perrydv** [comment]
yes that's it.  I'll run tests too, and on Windows tomorrow.  Want to make sure the compiler flags from the other issue don't trigger other problems on systems where they weren't needed.
